xdem/coreg/affine.py

```diff
@@ -54,147 +54,728 @@
+            f"your DEMs with DEM.reproject() in a local projected CRS such as UTM, that you can find "
+            f"using DEM.get_metric_crs()."
+        )
+
+    # First, derive auxiliary variables of Nuth and Kääb (slope tangent, and aspect) for any point-raster input
+    slope_tan, aspect = _nuth_kaab_aux_vars(ref_elev=ref_elev, tba_elev=tba_elev)
+
+    # Add slope tangents near zero to outliers, to avoid infinite values from later division by slope tangent, and to
+    # subsample the right number of subsample points straight ahead
+    mask_zero_slope_tan = np.isclose(slope_tan, 0)
+    slope_tan[mask_zero_slope_tan] = np.nan
+
+    # Then, perform preprocessing: subsampling and interpolation of inputs and auxiliary vars at same points
+    aux_vars = {"slope_tan": slope_tan, "aspect": aspect}  # Wrap auxiliary data in dictionary to use generic function
+    sub_dh_interpolator, sub_aux_vars, subsample_final = _preprocess_pts_rst_subsample_interpolator(
+        params_random=params_random,
+        ref_elev=ref_elev,
+        tba_elev=tba_elev,
+        inlier_mask=inlier_mask,
+        aux_vars=aux_vars,
+        transform=transform,
+        area_or_point=area_or_point,
+        verbose=verbose,
+        z_name=z_name,
+    )
+
+    if verbose:
+        print("   Iteratively estimating horizontal shift:")
+    # Initialise east, north and vertical offset variables (these will be incremented up and down)
+    initial_offset = (0.0, 0.0, 0.0)
+    # Resolution
+    res = _res(transform)
+    # Iterate through method of Nuth and Kääb (2011) until tolerance or max number of iterations is reached
+    assert sub_aux_vars is not None  # Mypy: dictionary cannot be None here
+    constant_inputs = (sub_dh_interpolator, sub_aux_vars["slope_tan"], sub_aux_vars["aspect"], res, params_fit_or_bin)
+    final_offsets = _iterate_method(
+        method=_nuth_kaab_iteration_step,
+        iterating_input=initial_offset,
+        constant_inputs=constant_inputs,
+        tolerance=tolerance,
+        max_iterations=max_iterations,
+        verbose=verbose,
+    )
+
+    return final_offsets, subsample_final
+
+
+########################
+# 2/ Gradient descending
+########################
 
-    # Estimate the a, b, and c parameters with least square minimisation
-    results = scipy.optimize.least_squares(
-        fun=residuals, x0=initial_guess, args=(y_medians, slice_bounds), xtol=1e-8, gtol=None, ftol=None
+
+def _gradient_descending_fit_func(
+    coords_offsets: tuple[float, float],
+    dh_interpolator: Callable[[float, float], NDArrayf],
+) -> float:
+    """
+    Fitting function of gradient descending method, returns the NMAD of elevation residuals.
+
+    :returns: NMAD of residuals.
+    """
+
+    # Calculate the elevation difference
+    dh = dh_interpolator(coords_offsets[0], coords_offsets[1])
+    vshift = -np.nanmedian(dh)
+    dh += vshift
```

Comment:
Removing the median bias should not change the value of the nmad, so these two lines are not needed I believe.

Reply:
Good point! I had just copied the original implementation blindly :sweat_smile: